catroid/src/main/java/org/catrobat/catroid/utils/CrashReporter.java

```diff
@@ -0,0 +1,134 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.utils;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+import android.support.annotation.VisibleForTesting;
+import android.util.Log;
+
+import com.crashlytics.android.Crashlytics;
+import com.google.gson.Gson;
+
+import org.catrobat.catroid.BuildConfig;
+import org.catrobat.catroid.ui.MainMenuActivity;
+import org.catrobat.catroid.ui.SettingsActivity;
+
+import io.fabric.sdk.android.Fabric;
+
+public final class CrashReporter {
+
+	private static final String TAG = CrashReporter.class.getSimpleName();
+
+	public static final String RECOVERED_FROM_CRASH = "RECOVERED_FROM_CRASH";
+	public static final String EXCEPTION_FOR_REPORT = "EXCEPTION_FOR_REPORT";
+
+	private static SharedPreferences preferences;
+	private static boolean isCrashReportEnabled = BuildConfig.CRASHLYTICS_CRASH_REPORT_ENABLED;
+
+	private CrashReporter() {
+	}
+
+	public static boolean initialize(Context context) {
+
+		preferences = PreferenceManager.getDefaultSharedPreferences(context);
+
+		if (isReportingEnabled()) {
+			Fabric.with(context, new Crashlytics());
+			Log.d(TAG, "INITIALIZED!");
+			return true;
+		}
+		Log.d(TAG, "INITIALIZATION FAILED! [ Anonymous Report : " + isAnonymousReportEnabled() + " Crash Report : " + isCrashReportEnabled + " ]");
+		return false;
+	}
+
+	private static boolean isAnonymousReportEnabled() {
+		return preferences != null && preferences.getBoolean(SettingsActivity.SETTINGS_CRASH_REPORTS, false);
+	}
+
+	private static boolean isRecoveredFromCrash() {
+		return preferences != null && preferences.getBoolean(RECOVERED_FROM_CRASH, false);
+	}
+
+	private static boolean isReportingEnabled() {
+		return isAnonymousReportEnabled() && isCrashReportEnabled;
+	}
+
+	@VisibleForTesting
+	public static void setIsCrashReportEnabled(boolean isEnabled) {
+		isCrashReportEnabled = isEnabled;
+	}
+
+	public static boolean logException(Throwable t) {
+
+		if (isReportingEnabled()) {
+			Crashlytics.logException(t);
+			return true;
+		}
+
+		return false;
+	}
+
+	public static boolean checkIfCrashRecoveryAndFinishActivity(final Activity context) {
+		Log.d(TAG, "AFTER_EXCEPTION : checkIfCrashRecoveryAndFinishActivity()");
+		if (isRecoveredFromCrash()) {
+			if (isReportingEnabled()) {
+				sendUnhandledCaughtException();
+			}
+			if (!(context instanceof MainMenuActivity)) {
+				context.finish();
+			} else {
+				preferences.edit().putBoolean(RECOVERED_FROM_CRASH, false).commit();
+				return true;
+			}
+		}
+		return false;
+	}
+
+	public static void sendUnhandledCaughtException() {
+		Log.d(TAG, "AFTER_EXCEPTION : sendCaughtException()");
+		Gson gson = new Gson();
+		String json = preferences.getString(EXCEPTION_FOR_REPORT, "");
+		Throwable exception = gson.fromJson(json, Throwable.class);
```

Comment:
@cheidenreich, all pull requests build on #2337. So you only need to review the last (or latest) commits, the first one can always be skipped. e.g. here you only need to review [pull/2365@b57e2d2](https://github.com/Catrobat/Catroid/pull/2365/commits/b57e2d285eea8bc768897a9e03a05c3c870ea97e) (commit may change and link may be broken in the future). Once #2337 is merged, we will do a rebase and remove the first commit from the PRs. We just need it now to build the project and fix all CrashReporter related tickets.

Yes, it will be handled in #2366.
This code is from #2337 where we kept as much code as possible from the original Firebase implementation. To create a better git history and make it easier to spot bugs later on.
In #2337 it isn't a problem because this code will only be executed if there is an exception stored and the SendReportDialog isn't currently used at all (we should remove it sometimes later on :wink:). In #2366 it definitely will be a problem :smile_cat:

But still, good catch!